### PR TITLE
IDENTITY-6529:Improving unit tests oidc component

### PR DESCRIPTION
**Pull request by GDRDABARERA**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java [hunk 1]
```diff
@@ -293,23 +470,72 @@ public void testSendRequest() throws Exception {
...
+    public void testGetOauthResponseWithExceptions() throws Exception {
+        OAuthClientRequest oAuthClientRequest = mock(OAuthClientRequest.class);
+        OAuthClient oAuthClient = mock(OAuthClient.class);
+        when(oAuthClient.accessToken(oAuthClientRequest)).thenThrow(OAuthSystemException.class);
+        setMockedLog(true);
+        openIDConnectAuthenticator.getOauthResponse(oAuthClient, oAuthClientRequest);
+    }
+
+    @Test
+    public void testGetOauthResponseWithOAuthProblemExceptions() throws Exception {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: No Asserts or Expected Exceptions?

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java [hunk 2]
```diff
@@ -293,23 +470,72 @@ public void testSendRequest() throws Exception {
         Assert.assertTrue(!result.isEmpty(), "The send request should not be empty.");
     }
 
+
+    @Test
+    public void testGetOauthResponseWithoutExceptions() throws Exception {
+        when(mockOAuthClient.accessToken(mockOAuthClientRequest)).thenReturn(mockOAuthJSONAccessTokenResponse);
+        openIDConnectAuthenticator.getOauthResponse(mockOAuthClient, mockOAuthClientRequest);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: No Asserts?

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java [hunk 3]
```diff
@@ -92,166 +142,296 @@
...
-    @Test(priority = 1)
+    @Test(dataProvider = "requestDataHandler")
+    public void testGetContextIdentifier(String grantType, String state, String loginType, String error, String debugEnable) throws Exception {
+        setMockedLog(Boolean.parseBoolean(debugEnable));
+        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE)).thenReturn(grantType);
+        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.OAUTH2_PARAM_STATE)).thenReturn(state);
+        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.LOGIN_TYPE)).thenReturn(loginType);
+        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.OAUTH2_ERROR)).thenReturn(error);
+
+        if (error != null) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: you can get the expected result from DataProvider and use `Assert.assertEquals(openIDConnectAuthenticator.getContextIdentifier(mockServletRequest), expected)`

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java [hunk 4]
```diff
@@ -92,166 +142,296 @@
             "oguftrCCGjBy1kKT4VqFLOqlA-8wUhOj_rZT9SUIBQRDPu0RZobvsskqYo40GEZrUoa";
     private static OAuthClientResponse token;
     private Map<String, String> paramValueMap;
+    private int TENAT_ID = 1234;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: Typo?

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java [hunk 5]
```diff
@@ -92,166 +142,296 @@
...
-        // When all parameters not null.
-        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE)).thenReturn("openid");
-        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.OAUTH2_PARAM_STATE)).thenReturn("active,OIDC");
-        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.LOGIN_TYPE)).thenReturn("BASIC");
-        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.OAUTH2_ERROR)).thenReturn("Error Login.");
-        Assert.assertTrue(openIDConnectAuthenticator.canHandle(mockServletRequest), "Invalid can handle response for the request.");
-        Assert.assertNotNull(openIDConnectAuthenticator.getContextIdentifier(mockServletRequest), "Invalid context identifier.");
+        return new String[][]{
+                {","},
+                {"|"}
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: Change this to something like ,,,

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java [hunk 6]
```diff
@@ -92,166 +135,264 @@
...
+    @Test(dataProvider = "requestDataHandler")
+    public void testCanHandle(String grantType, String state, String loginType, String error, String expectedCanHandler, String expectedContext, String msgCanHandler, String msgContext) throws Exception {
+
+
+        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE)).thenReturn(grantType);
+        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.OAUTH2_PARAM_STATE)).thenReturn(state);
+        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.LOGIN_TYPE)).thenReturn(loginType);
+        when(mockServletRequest.getParameter(OIDCAuthenticatorConstants.OAUTH2_ERROR)).thenReturn(error);
+
+        Assert.assertEquals(openIDConnectAuthenticator.canHandle(mockServletRequest), Boolean.parseBoolean(expectedCanHandler), msgCanHandler);
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: Static import Assert.* methods

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java [hunk 7]
```diff
@@ -92,166 +135,264 @@
...
-        when(mockOAuthClientResponse.getParam(OIDCAuthenticatorConstants.ACCESS_TOKEN)).thenReturn("eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9");
+        // Test with no json response.
+        when(mockOAuthClientResponse.getParam(OIDCAuthenticatorConstants.ACCESS_TOKEN)).
+                thenReturn("eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9");
+        result = openIDConnectAuthenticator.getSubjectAttributes(mockOAuthClientResponse, authenticatorProperties);
+        Assert.assertTrue(result.isEmpty(), "result is not Empty.");
+
+        // Test with a json response which is not empty.
+        Map<String, Object> jsonObject = new HashMap<>();
+        jsonObject.put("email", new String("{\"www.wso2.org/email\" : \"dinali@wso2.com\"}"));
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: avoid real emails

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java [hunk 8]
```diff
@@ -51,33 +63,64 @@
...
+import static org.powermock.api.mockito.PowerMockito.doReturn;
+import static org.powermock.api.mockito.PowerMockito.mock;
 import static org.powermock.api.mockito.PowerMockito.mockStatic;
 import static org.powermock.api.mockito.PowerMockito.when;
 import static org.powermock.api.mockito.PowerMockito.whenNew;
 
 @PrepareForTest({LogFactory.class, OAuthClient.class, URL.class, FrameworkUtils.class, OpenIDConnectAuthenticatorServiceComponent.class,
-        OAuthAuthzResponse.class, OAuthClientRequest.class, OAuthClientResponse.class, OAuthClientRequest.AuthenticationRequestBuilder.class})
+        OAuthAuthzResponse.class, OAuthClientRequest.class, OAuthClientResponse.class, IdentityUtil.class})
 public class OpenIDConnectAuthenticatorTest {
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: Please add a class definition

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java [hunk 9]
```diff
@@ -293,23 +430,69 @@ public void testSendRequest() throws Exception {
         Assert.assertTrue(!result.isEmpty(), "The send request should not be empty.");
     }
 
+
+    @Test
+    public void testGetOauthResponseWithoutExceptions() throws Exception {
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: Why did we throw a generic exception

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java (hunk 1 above)
Comment (DMHP) [reply components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: Also why did we throw a generic exception?

Comment (GDRDABARERA) [reply components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: done

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java (hunk 9 above)
Comment (GDRDABARERA) [reply components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: done

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java (hunk 6 above)
Comment (GDRDABARERA) [reply components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: done

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java (hunk 8 above)
Comment (GDRDABARERA) [reply components/org.wso2.carbon.identity.application.authenticator.oidc/src/test/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticatorTest.java]: done

**Pull request merged**